File: cpp/cpp_1000-/1282.cc
// C++
// leetcode 1282
// https://leetcode.cn/problems/group-the-people-given-the-group-size-they-belong-to/

class Solution {
public:
    vector<vector<int>> groupThePeople(vector<int>& groupSizes) {
    	map<int, vector<int>> m;
    	int n = groupSizes.size();
    	for (int i = 0; i < n; i++) {
    		m[groupSizes[i]].push_back(i);
    	}

    	vector<vector<int>> res;

    	for (auto &[k, v] : m) {
    		vector<int> tmp;
    		for (int i = 0; i < v.size(); i++) {
    			tmp.push_back(v[i]);
    			if (tmp.size() == k) {
    				res.push_back(tmp);
    				tmp.clear();
    			}
    		}
    	}

    	return res;
    }
};

File: Go/Go_1000-/1282.go
// Go
// leetcode 1282
// https://leetcode.cn/problems/group-the-people-given-the-group-size-they-belong-to/

func groupThePeople(groupSizes []int) (res [][]int) {
	m := make(map[int][]int)
	for i, v := range groupSizes {
		if len(m[v]) == 0 || len(m[v]) == v {
			m[v] = make([]int, 0, v)
		}
		m[v] = append(m[v], i)
		if len(m[v]) == v {
			res = append(res, m[v])
		}
	}
	return res
}